### PR TITLE
VRD-386 Promote self contained builds

### PR DESCRIPTION
**Pull request by jared-verta**
## Impact and Context

This PR changes the Promotion script to directly promoting a specific build of a model version, as opposed to promoting an endpoint and its associated build.

- Environment variable configuration has changed:
  - VERTA_SOURCE_MODEL_VERSION_ID added as the configuration parameter to specify the source of promotion
  - VERTA_SOURCE_ENDPOINT_PATH removed
  - VERTA_DEST_ENDPOINT_PATH removed
- Fixed duplicate upload of `model.pkl` artifact
- Find the latest self contained build of the promoted model version
- Warn and exit if no candidate builds are found
- Added a stub function `upload_build` which can be used as an extension point to add custom upload behavior for S3 artifacts. Input to the function is a dict which contains an S3 path in the `location` key. The output of this function must be the path of the uploaded build in the destination S3 bucket. If the promotion source and destinations share a common S3 bucket/access control, this function will work as is.

This PR also adds a utility script to facilitate creation of self contained builds.

### CODE REVIEW

File: deployment/promotion/promote.py [hunk 1]
```diff
@@ -227,110 +227,83 @@ def upload_artifact(auth, model_version_id, artifact):
...
 
-    promotion_data = {
-        'endpoint': endpoint,
-        'stages': []
+    time_format = '%Y-%m-%dT%H:%M:%S.%fZ'
+
+    model_version_builds = []
+    for build in all_builds['builds']:
+        if 'model_version_id' in build['creator_request'] and 'self_contained' in build['creator_request']:
+            if build['creator_request']['model_version_id'] == model_version_id:
```
Comment (hmacdonald-verta) [inline deployment/promotion/promote.py:254]: You already restricted by MV ID in the `get_builds` function, so why are you checking again

Comment (hmacdonald-verta) [reply deployment/promotion/promote.py:254]: I think you can get rid of this check and move the "no self contained builds found" check up above this loop, then handle just timestamps in this loop.

Comment (jared-verta) [reply deployment/promotion/promote.py:254]: You're right that the model version id check is redundant, I'll move that. But we won't know if there are any self contained builds or not until the loop is complete. Can also embed the date check into the loop as well

File: deployment/promotion/promote.py [hunk 2]
```diff
@@ -248,27 +248,24 @@ def get_promotion_data(_config):
...
-    for mv_build in model_version_builds:
-        if mv_build['parsed_date'] > build['parsed_date']:
-            build = mv_build
-
     model = get_registered_model(source_auth,  model_version['registered_model_id'])
     artifacts = download_artifacts(source_auth, model_version_id, model_version['artifacts'], model_version['model'])
 
     promotion = {
-         'build': build,
+         'build': b,
```
Comment (hmacdonald-verta) [inline deployment/promotion/promote.py:268]: I think this should still be `build`? I don't see `b` in scope here (python letting you do sketchy stuff lol)

Comment (jared-verta) [reply deployment/promotion/promote.py:268]: Ah yep, stupid pycharm thanks!

**Review by hmacdonald-verta — APPROVED** [review]

**Pull request merged**